core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java

```diff
@@ -413,6 +414,30 @@ public void writeMap(@Nullable Map<String, Object> map) throws IOException {
     }
 
     /**
+     * write map to stream with consistent order
+     * to make sure every map generated bytes order are same.
+     * This method is compatible with {@code StreamInput.readMap} and {@code StreamInput.readGenericValue}
+     * This method only will handle the map keys order, not maps contained within the map
+     */
+    public void writeMapWithConsistentOrder(@Nullable Map<String, ? extends Object> map)
+        throws IOException {
+        if (map == null) {
+            writeByte((byte) -1);
+            return;
+        }
+        assert false == (map instanceof LinkedHashMap);
```

Comment:
I think it is fairly safe as an assertion. Mostly we are in control of the types we pass in here. Any `LinkedHashMap` isn't *wrong* to pass here, it is just super silly because we'll reorder it. If we want something stronger than an assertion then maybe we should do something like
```java
if (map instanceof LinkedHashMap) {
    // Already has consistent order
    writeMap(map);
    return;
}
```
?